source/common/buffer/buffer_impl.cc

```diff
@@ -277,6 +277,30 @@ void* OwnedImpl::linearize(uint32_t size) {
   return slices_.front()->data();
 }
 
+RawSlice OwnedImpl::maybeLinearize(uint32_t max_size, uint32_t desired_min_size) {
+  while (!slices_.empty() && slices_[0]->dataSize() == 0) {
+    slices_.pop_front();
+  }
+
+  if (slices_.empty()) {
+    return {nullptr, 0};
+  }
+
+  const uint64_t slice_size = std::min<uint64_t>(slices_[0]->dataSize(), max_size);
+  if (slice_size >= desired_min_size) {
+    return {slices_[0]->data(), slice_size};
+  }
+
+  // The next slice will already be of the desired size, so don't copy and
+  // return the front slice.
```

Comment:
There's an error in the next line, should be:
if (slices_.size() >= 2 && slices_[1]->dataSize() >= desired_min_size) {

Reply:
Given that it's a heuristic, I wouldn't say it's an error, just a choice. If the next slice is slightly larger than 4k, and the current slice is 1 byte, what's the best behavior?

It turns out the slice sizes are terrible, as you've noted, due to the inline storage of the OwnedSlice. The second slice contains just slightly less than 16k (I think it's 64 bytes less), which results in a bunch of copies on subsequent slices.

I think the next step is to remove the inline-storage from the slice (#14111), then re-evaluate this PR.

Reply:
Sorry, my read of the comment made me think that you intended to compare against desired_min_size.

I think an interesting case is write behavior for HTTP2 which involves a 9 byte data frame header followed by up to 16kb of data.  The change in #14111 will have some consequences to how said writes interact with linearize, but I think would result in little to no performance consequences since both the versions of the buffer class would end up copying about the same amount of data during linearize.

Reply:
> Sorry, my read of the comment made me think that you intended to compare against desired_min_size.

I see the confusion now. That comment was written when the parameter had a different (less clear) name. I'll clarify the comment.